### PR TITLE
Moby build operation and linuxkit push

### PR DESCRIPTION
**Pull request by amegianeg**
I've cleaned some code to use Moby's structs and methods (get a Moby config from a linuxkit file, use it to build the image).
There is an example file in `example/example.go` that can be executed to see the output of the flow (will implement a test for this when finished). It fails because there is no trust data for the Nginx image we are using. Here is the traceback of the example.go file execution:
```
alberto-tutum:i2kit alberto$ go run example/example.go
INFO[0000] Extract kernel image: linuxkit/kernel:4.9.50
INFO[0001] Pull image: docker.io/linuxkit/kernel:4.9.50@sha256:629d6ae93b7635caf0e46e7d4520463dd959b4f0b185c0afd6ba4997dd13089a
INFO[0014] Add init containers:
INFO[0014] Process init image: linuxkit/init:851e9c3ad0574d640b733b92fdb26c368d2f7f8f
INFO[0015] Pull image: docker.io/linuxkit/init:851e9c3ad0574d640b733b92fdb26c368d2f7f8f@sha256:359ee2eeeecf22aa4898ae3085d7ec6d615b01198e89043cbbe96078afe7ef85
INFO[0022] Process init image: linuxkit/runc:a1b564248a0d0b118c11e61db9f84ecf41dd2d2a
INFO[0023] Pull image: docker.io/linuxkit/runc:a1b564248a0d0b118c11e61db9f84ecf41dd2d2a@sha256:ebca5a92cafa4436e0e49fc3c526350940c7a0c301978e7bd8b2b65e2b139e33
INFO[0025] Process init image: linuxkit/containerd:06876ceef325e49e9ba119659357768d5df89075
INFO[0026] Pull image: docker.io/linuxkit/containerd:06876ceef325e49e9ba119659357768d5df89075@sha256:c02c1e1fbbf7efeeba437423a4d13064fb7c9082d5e8d2f3d648740464c31637
INFO[0029] Process init image: linuxkit/ca-certificates:e44b0a66df5a102c0e220f0066b0d904710dcb10
INFO[0030] Pull image: docker.io/linuxkit/ca-certificates:e44b0a66df5a102c0e220f0066b0d904710dcb10@sha256:1176e802e26c3774d8f3c318f495b53bf143ec954efcd02e0ff7e6769960b0cd
INFO[0032] Add onboot containers:
INFO[0032]   Create OCI config for linuxkit/sysctl:154913b72c6f1f33eb408609fca9963628e8c051
INFO[0033] Pull image: docker.io/linuxkit/sysctl:154913b72c6f1f33eb408609fca9963628e8c051@sha256:a2e866e3b3b765772c71678a8dd3722bc32039378e330ed22ea224e42268430b
INFO[0036]   Create OCI config for linuxkit/dhcpcd:f3f5413abb78fae9020e35bd4788fa93df4530b7
INFO[0037] Pull image: docker.io/linuxkit/dhcpcd:f3f5413abb78fae9020e35bd4788fa93df4530b7@sha256:3f5cafe744b719ab45f1515174387859d647f387dc79b0956e8a983ed04d3da4
INFO[0039]   Create OCI config for linuxkit/metadata:da3138079c168e0c5608d8f3853366c113ed91d2
INFO[0041] Pull image: docker.io/linuxkit/metadata:da3138079c168e0c5608d8f3853366c113ed91d2@sha256:ec0de477af801e0c7301927a8bdff01be484745f78722142532cd74bf5faf23b
INFO[0043] Add service containers:
INFO[0043]   Create OCI config for linuxkit/rngd:558e86a36242bb74353bc9287b715ddb8567357e
INFO[0044] Pull image: docker.io/linuxkit/rngd:558e86a36242bb74353bc9287b715ddb8567357e@sha256:4e74cd3e35b69589e6c91bc6cabd7994f2d4f943b438418edef0674662fc7e96
INFO[0046]   Create OCI config for nginx:1.7.9
Failed to create OCI spec for nginx:1.7.9: Trusted pull for nginx:1.7.9 failed: No trust data for 1.7.9
```
But starts to look good :)
After it, will implement linuxkit pushAWS (https://github.com/linuxkit/linuxkit/blob/master/src/cmd/linuxkit/push_aws.go). I haven't found any importable code for it, will try to dig deeper, but if I don't see it, will use linuxkit tooling.

### CODE REVIEW

File: linuxkit/io.go [hunk 1]
```diff
@@ -29,19 +29,6 @@ type trustYml struct {
...
-	}
-	var result templateYml
-	err = yaml.Unmarshal(ymlBytes, &result)
-	if err != nil {
-		return nil, err
-	}
-	return &result, nil
-}
-
 func write(t *templateYml, path string) error {
```
Comment (pchico83) [inline linuxkit/io.go]: I would remove this and inline it where it is used.

Comment (pchico83) [reply linuxkit/io.go]: Structs in this file are not longer needed, right?

**Review by pchico83 — APPROVED** [review]
LGMT! Thanks for taking care of this.
Comments:
- I would remove `io.go`.
- I would remove `example.go`. We can have a deploy command flag to just print the CF template but without deploying it. This flag will be helpful for use (to avoid deployments when testing, also for unit tests) and for the user, to know what is gonna be deployed before spending money.

**Comment by amegianeg** [comment]
Sorry @pchico83 I forgot to remove `io.go` and `io_test.go` files! updated.
For the testing flag -> https://github.com/pchico83/i2kit/issues/3

**Comment by amegianeg** [comment]
Changing the version of Nginx worked for the build
```
alberto-tutum:i2kit alberto$ go run example/example.go
INFO[0000] Extract kernel image: linuxkit/kernel:4.9.50
INFO[0001] Pull image: docker.io/linuxkit/kernel:4.9.50@sha256:629d6ae93b7635caf0e46e7d4520463dd959b4f0b185c0afd6ba4997dd13089a
INFO[0005] Add init containers:
INFO[0005] Process init image: linuxkit/init:851e9c3ad0574d640b733b92fdb26c368d2f7f8f
INFO[0007] Pull image: docker.io/linuxkit/init:851e9c3ad0574d640b733b92fdb26c368d2f7f8f@sha256:359ee2eeeecf22aa4898ae3085d7ec6d615b01198e89043cbbe96078afe7ef85
INFO[0009] Process init image: linuxkit/runc:a1b564248a0d0b118c11e61db9f84ecf41dd2d2a
INFO[0011] Pull image: docker.io/linuxkit/runc:a1b564248a0d0b118c11e61db9f84ecf41dd2d2a@sha256:ebca5a92cafa4436e0e49fc3c526350940c7a0c301978e7bd8b2b65e2b139e33
INFO[0014] Process init image: linuxkit/containerd:06876ceef325e49e9ba119659357768d5df89075
INFO[0015] Pull image: docker.io/linuxkit/containerd:06876ceef325e49e9ba119659357768d5df89075@sha256:c02c1e1fbbf7efeeba437423a4d13064fb7c9082d5e8d2f3d648740464c31637
INFO[0019] Process init image: linuxkit/ca-certificates:e44b0a66df5a102c0e220f0066b0d904710dcb10
INFO[0020] Pull image: docker.io/linuxkit/ca-certificates:e44b0a66df5a102c0e220f0066b0d904710dcb10@sha256:1176e802e26c3774d8f3c318f495b53bf143ec954efcd02e0ff7e6769960b0cd
INFO[0023] Add onboot containers:
INFO[0023]   Create OCI config for linuxkit/sysctl:154913b72c6f1f33eb408609fca9963628e8c051
INFO[0024] Pull image: docker.io/linuxkit/sysctl:154913b72c6f1f33eb408609fca9963628e8c051@sha256:a2e866e3b3b765772c71678a8dd3722bc32039378e330ed22ea224e42268430b
INFO[0026]   Create OCI config for linuxkit/dhcpcd:f3f5413abb78fae9020e35bd4788fa93df4530b7
INFO[0027] Pull image: docker.io/linuxkit/dhcpcd:f3f5413abb78fae9020e35bd4788fa93df4530b7@sha256:3f5cafe744b719ab45f1515174387859d647f387dc79b0956e8a983ed04d3da4
INFO[0029]   Create OCI config for linuxkit/metadata:da3138079c168e0c5608d8f3853366c113ed91d2
INFO[0030] Pull image: docker.io/linuxkit/metadata:da3138079c168e0c5608d8f3853366c113ed91d2@sha256:ec0de477af801e0c7301927a8bdff01be484745f78722142532cd74bf5faf23b
INFO[0033] Add service containers:
INFO[0033]   Create OCI config for linuxkit/rngd:558e86a36242bb74353bc9287b715ddb8567357e
INFO[0034] Pull image: docker.io/linuxkit/rngd:558e86a36242bb74353bc9287b715ddb8567357e@sha256:4e74cd3e35b69589e6c91bc6cabd7994f2d4f943b438418edef0674662fc7e96
INFO[0036]   Create OCI config for nginx:1.12.1
INFO[0037] Pull image: docker.io/library/nginx:1.12.1@sha256:734eace1881a0d976efa9fbe7edec9accac60dd8878327d8f769334e7bd637ae
INFO[0051] Pull image: docker.io/library/nginx:1.12.1@sha256:734eace1881a0d976efa9fbe7edec9accac60dd8878327d8f769334e7bd637ae
INFO[0054]   Create OCI config for nginx:1.12.1
```
(from https://github.com/moby/tool/blob/b9de5ff4c64e01d8fbaa495776890a82e9647f20/src/moby/docker.go#L109)

**Comment by amegianeg** [comment]
@pchico83 I have updated this PR, the image file wasn't created for AWS needs (raw). Please review

**Review by pchico83 — APPROVED** [review]
Cool!

**Comment by amegianeg** [comment]
Added `linuxkit` push to AWS and convert to AMI.

Needs:
`AWS_SHARED_CREDENTIALS_FILE` pointing to AWS file credentials (AWS_CREDENTIALS).
Linuxkit installed (haven't found _yet_ the way to use it in the code): `go get -u github.com/linuxkit/linuxkit/src/cmd/linuxkit`
`AWS_REGION` environment variable as this is used by the AWS Go SDK.

NOTE: error handling needs review.
Linuxkit is used as a binary, haven't found the way to use it as a library (yet)

**Comment by amegianeg** [comment]
This is the output of an example deploy (fails because we have hardcoded data like SSH keys and subnets that are not valid, but you get an idea of how it works):

```
alberto-tutum:i2kit alberto$ go run main.go deploy
INFO[0000] >>> Moby build started...
INFO[0000] Extract kernel image: linuxkit/kernel:4.9.50
INFO[0002] Pull image: docker.io/linuxkit/kernel:4.9.50@sha256:629d6ae93b7635caf0e46e7d4520463dd959b4f0b185c0afd6ba4997dd13089a
INFO[0005] Add init containers:
INFO[0005] Process init image: linuxkit/init:851e9c3ad0574d640b733b92fdb26c368d2f7f8f
INFO[0006] Pull image: docker.io/linuxkit/init:851e9c3ad0574d640b733b92fdb26c368d2f7f8f@sha256:359ee2eeeecf22aa4898ae3085d7ec6d615b01198e89043cbbe96078afe7ef85
INFO[0008] Process init image: linuxkit/runc:a1b564248a0d0b118c11e61db9f84ecf41dd2d2a
INFO[0009] Pull image: docker.io/linuxkit/runc:a1b564248a0d0b118c11e61db9f84ecf41dd2d2a@sha256:ebca5a92cafa4436e0e49fc3c526350940c7a0c301978e7bd8b2b65e2b139e33
INFO[0012] Process init image: linuxkit/containerd:06876ceef325e49e9ba119659357768d5df89075
INFO[0013] Pull image: docker.io/linuxkit/containerd:06876ceef325e49e9ba119659357768d5df89075@sha256:c02c1e1fbbf7efeeba437423a4d13064fb7c9082d5e8d2f3d648740464c31637
INFO[0016] Process init image: linuxkit/ca-certificates:e44b0a66df5a102c0e220f0066b0d904710dcb10
INFO[0017] Pull image: docker.io/linuxkit/ca-certificates:e44b0a66df5a102c0e220f0066b0d904710dcb10@sha256:1176e802e26c3774d8f3c318f495b53bf143ec954efcd02e0ff7e6769960b0cd
INFO[0019] Add onboot containers:
INFO[0019]   Create OCI config for linuxkit/sysctl:154913b72c6f1f33eb408609fca9963628e8c051
INFO[0021] Pull image: docker.io/linuxkit/sysctl:154913b72c6f1f33eb408609fca9963628e8c051@sha256:a2e866e3b3b765772c71678a8dd3722bc32039378e330ed22ea224e42268430b
INFO[0023]   Create OCI config for linuxkit/dhcpcd:f3f5413abb78fae9020e35bd4788fa93df4530b7
INFO[0024] Pull image: docker.io/linuxkit/dhcpcd:f3f5413abb78fae9020e35bd4788fa93df4530b7@sha256:3f5cafe744b719ab45f1515174387859d647f387dc79b0956e8a983ed04d3da4
INFO[0026]   Create OCI config for linuxkit/metadata:da3138079c168e0c5608d8f3853366c113ed91d2
INFO[0027] Pull image: docker.io/linuxkit/metadata:da3138079c168e0c5608d8f3853366c113ed91d2@sha256:ec0de477af801e0c7301927a8bdff01be484745f78722142532cd74bf5faf23b
INFO[0029] Add service containers:
INFO[0029]   Create OCI config for linuxkit/rngd:558e86a36242bb74353bc9287b715ddb8567357e
INFO[0030] Pull image: docker.io/linuxkit/rngd:558e86a36242bb74353bc9287b715ddb8567357e@sha256:4e74cd3e35b69589e6c91bc6cabd7994f2d4f943b438418edef0674662fc7e96
INFO[0032]   Create OCI config for nginx:1.12.1
INFO[0033] Pull image: docker.io/library/nginx:1.12.1@sha256:734eace1881a0d976efa9fbe7edec9accac60dd8878327d8f769334e7bd637ae
INFO[0037]   Create OCI config for nginx:1.12.1
INFO[0037] Create outputs in /Users/alberto/.moby/nginx-deployment:
INFO[0037]   /Users/alberto/.moby/nginx-deployment.raw
INFO[0091] >>> Linuxkit run started... (this process will take a while)
linuxkit err | time="2017-09-22T13:08:02-07:00" level=debug msg="PutObject:
linuxkit err | {
linuxkit err |   Body: buffer(0xc42000e0e8),
linuxkit err |   Bucket: \"linuxkit\",
linuxkit err |   ContentLength: 1073844224,
linuxkit err |   ContentType: \"application/octet-stream\",
linuxkit err |   Key: \"nginx-deployment.raw\"
linuxkit err | }"
linuxkit err | time="2017-09-22T13:26:55-07:00" level=debug msg="ImportSnapshot:
linuxkit err | {
linuxkit err |   Description: \"LinuxKit: nginx-deployment\",
linuxkit err |   DiskContainer: {
linuxkit err |     Description: \"LinuxKit: nginx-deployment disk\",
linuxkit err |     Format: \"raw\",
linuxkit err |     UserBucket: {
linuxkit err |       S3Bucket: \"linuxkit\",
linuxkit err |       S3Key: \"nginx-deployment.raw\"
linuxkit err |     }
linuxkit err |   }
linuxkit err | }"
linuxkit err | time="2017-09-22T13:26:57-07:00" level=debug msg="DescribeImportSnapshotTask:
linuxkit err | {
linuxkit err |   ImportTaskIds: [\"import-snap-fgb39ov7\"]
linuxkit err | }"
linuxkit err | time="2017-09-22T13:26:57-07:00" level=debug msg="Task import-snap-fgb39ov7 is 3% complete. Waiting 60 seconds...
linuxkit err | "
linuxkit err | time="2017-09-22T13:27:57-07:00" level=debug msg="DescribeImportSnapshotTask:
linuxkit err | {
linuxkit err |   ImportTaskIds: [\"import-snap-fgb39ov7\"]
linuxkit err | }"
linuxkit err | time="2017-09-22T13:27:58-07:00" level=debug msg="Task import-snap-fgb39ov7 is 5% complete. Waiting 60 seconds...
linuxkit err | "
linuxkit err | time="2017-09-22T13:28:58-07:00" level=debug msg="DescribeImportSnapshotTask:
linuxkit err | {
linuxkit err |   ImportTaskIds: [\"import-snap-fgb39ov7\"]
linuxkit err | }"
linuxkit err | time="2017-09-22T13:28:59-07:00" level=debug msg="Task import-snap-fgb39ov7 is 43% complete. Waiting 60 seconds...
linuxkit err | "
linuxkit err | time="2017-09-22T13:29:59-07:00" level=debug msg="DescribeImportSnapshotTask:
linuxkit err | {
linuxkit err |   ImportTaskIds: [\"import-snap-fgb39ov7\"]
linuxkit err | }"
linuxkit err | time="2017-09-22T13:30:00-07:00" level=debug msg="Task import-snap-fgb39ov7 is 92% complete. Waiting 60 seconds...
linuxkit err | "
linuxkit err | time="2017-09-22T13:31:00-07:00" level=debug msg="DescribeImportSnapshotTask:
linuxkit err | {
linuxkit err |   ImportTaskIds: [\"import-snap-fgb39ov7\"]
linuxkit err | }"
linuxkit err | time="2017-09-22T13:31:01-07:00" level=debug msg="SnapshotID: %!s(*string=0xc42001abc8)"
linuxkit err | time="2017-09-22T13:31:01-07:00" level=debug msg="RegisterImage:
linuxkit err | {
linuxkit err |   Architecture: \"x86_64\",
linuxkit err |   BlockDeviceMappings: [{
linuxkit err |       DeviceName: \"/dev/sda1\",
linuxkit err |       Ebs: {
linuxkit err |         DeleteOnTermination: true,
linuxkit err |         SnapshotId: \"snap-0072bd46731d160af\",
linuxkit err |         VolumeType: \"standard\"
linuxkit err |       }
linuxkit err |     }],
linuxkit err |   Description: \"LinuxKit: nginx-deployment image\",
linuxkit err |   Name: \"nginx-deployment\",
linuxkit err |   RootDeviceName: \"/dev/sda1\",
linuxkit err |   VirtualizationType: \"hvm\"
linuxkit err | }"
linuxkit err | time="2017-09-22T13:31:01-07:00" level=info msg="Created AMI: ami-ce59e8a1"
2017-09-22 13:31:02: nginx-deployment (AWS::CloudFormation::Stack) CREATE_IN_PROGRESS User Initiated
2017-09-22 13:31:06: ELB (AWS::ElasticLoadBalancing::LoadBalancer)  CREATE_IN_PROGRESS
2017-09-22 13:31:07: LaunchConfig (AWS::AutoScaling::LaunchConfiguration)  CREATE_IN_PROGRESS
2017-09-22 13:31:07: LaunchConfig (AWS::AutoScaling::LaunchConfiguration) CREATE_IN_PROGRESS Resource creation Initiated
2017-09-22 13:31:07: LaunchConfig (AWS::AutoScaling::LaunchConfiguration)  CREATE_COMPLETE
2017-09-22 13:31:07: ELB (AWS::ElasticLoadBalancing::LoadBalancer) CREATE_FAILED InvalidSubnetID.NotFound
2017-09-22 13:31:09: nginx-deployment (AWS::CloudFormation::Stack) ROLLBACK_IN_PROGRESS The following resource(s) failed to create: [ELB]. . Rollback requested by user.
```
And the template:
```
{
    "AWSTemplateFormatVersion": "2010-09-09",
    "Resources": {
        "ASG": {
            "Type": "AWS::AutoScaling::AutoScalingGroup",
            "Properties": {
                "HealthCheckGracePeriod": 120,
                "LaunchConfigurationName": {
                    "Ref": "LaunchConfig"
                },
                "LoadBalancerNames": [
                    {
                        "Ref": "ELB"
                    }
                ],
                "MaxSize": "3",
                "MinSize": "3",
                "VPCZoneIdentifier": [
                    "subnet-3f087e57"
                ]
            }
        },
        "DNSRecord": {
            "Type": "AWS::Route53::RecordSet",
            "Properties": {
                "HostedZoneName": "i2kit.io.",
                "Name": "nginx-deployment.i2kit.io.",
                "ResourceRecords": [
                    {
                        "Fn::GetAtt": [
                            "ELB",
                            "DNSName"
                        ]
                    }
                ],
                "TTL": "900",
                "Type": "CNAME"
            }
        },
        "ELB": {
            "Type": "AWS::ElasticLoadBalancing::LoadBalancer",
            "Properties": {
                "LoadBalancerName": "testing-i2kit",
                "Listeners": [
                    {
                        "InstancePort": "80",
                        "InstanceProtocol": "HTTP",
                        "LoadBalancerPort": "80",
                        "Protocol": "HTTP"
                    },
                    {
                        "InstancePort": "90",
                        "InstanceProtocol": "HTTP",
                        "LoadBalancerPort": "90",
                        "Protocol": "HTTP"
                    }
                ],
                "Subnets": [
                    "hello"
                ]
            }
        },
        "LaunchConfig": {
            "Type": "AWS::AutoScaling::LaunchConfiguration",
            "Properties": {
                "ImageId": "ami-ce59e8a1",
                "InstanceType": "t2.micro",
                "KeyName": "pablo",
                "SecurityGroups": [
                    "sg-5d42b836"
                ]
            }
        }
    },
    "Outputs": {
        "URL": {
            "Description": "The URL of the stack",
            "Value": {
                "Fn::Join": [
                    "",
                    [
                        "http://",
                        {
                            "Fn::GetAtt": [
                                "ELB",
                                "DNSName"
                            ]
                        }
                    ]
                ]
            }
        }
    }
}
```

**Comment by amegianeg** [comment]
BTW, do not know why `linuxkit` is sending logs to stderr instead of stdout

**Pull request merged**